### ISSUE TITLE
incorrect weights in scatter plots when slicing or masking

### ISSUE THREAD

**Issue by lukashergt** [williamjameshandley/anesthetic#95]
When plotting a scatter plot of a masked or sliced `NestedSamples` instance, the weights seem to be incorrect. Could this be a normalisation issue (if so, worth to keep in mind for #90)? 

The data used in the MWE contains some `-inf` entries in `logL`, which are excluded from any plotting. Hence, masking all these `-inf` entries should have no effect on a resulting plot. This seems to be correct for `'kde'` plots, but wrong for `'scatter'` plots. In the MWE below, blue refers to the masked data and orange to the full data, which are not the same in the scatter plots.

This issue becomes very clear when looking at how slicing and masking affects the axes limit behaviour. Compare for example the axis range for `r` in the MWE here to the MWE in #94.

### MWE

data: [unilog_r.log](https://github.com/williamjameshandley/anesthetic/files/4746501/unilog_r.log) (same data as in #93 and #94)

```python
data = pandas.read_csv("/home/hergtl/data/AnestheticData/unilog_r.log", sep='\t')
ns = NestedSamples(data=data[['n_s', 'r', 'logr']], logL=data.logL, logL_birth=data.logL_birth, w=data.weight)
mask = np.isfinite(ns.logL)
fig = plt.figure(figsize=(5, 5))
fig, axes = make_2d_axes(['n_s', 'r', 'logr'], fig=fig)
ns[mask].plot_2d(axes, label='masked')
ns.plot_2d(axes, label='full')
axes.iloc[-1, 0].legend(bbox_to_anchor=(len(axes), len(axes)), loc='upper left');
```

To look at slicing instead, simply change `[mask]` to e.g. `[412:]`:
```python
ns[412:].plot_2d(axes, label='sliced')
```

![unilog_full_vs_masked](https://user-images.githubusercontent.com/18258042/84041286-8a16da80-a99b-11ea-9d15-c995e9154235.png)

**Additional context**

Note that this issue goes beyond `-inf` entries in `logL` but applies more generally to masking and slicing.

**Comment by lukashergt** [comment]
### Simpler MWE:

```python
ns = NestedSamples(root="anesthetic/tests/example_data/pc")
fig, axes = make_2d_axes(['x0', 'x1'])
ns.plot_2d(axes)
ns[1:].plot_2d(axes);
```

![full_vs_sliced](https://user-images.githubusercontent.com/18258042/84073700-33c09080-a9c9-11ea-8210-62844510e519.png)

This highlights how the removal of a single sample point makes the scatter plot blow up.

**Comment by lukashergt** [comment]
We can further simplify the problem by comparing

```python
ns['x0'].weight
```

with 

```python
ns[1:]['x0'].weight
```

Without slicing we get the correct weights. With slicing we get an array of ones.

Seems like this is linked to the treatment of weights in anesthetic as described in #86.

**Comment by lukashergt** [comment]
A fairly easy fix would be to always pass the weights directly to the `.compress` method, i.e. change line 205 in samples.py to:
```diff
-   x = self[paramname_x].compress(ncompress)
+   x = self[paramname_x].compress(weight=self.weight, nsamples=ncompress)
```

However, this goes a bit against the spirit of the `WeightedSeries` and `WeightedDataFrame` classes. 

@williamjameshandley: Would you like me to create a PR for this fix or would you prefer to defer this to a more fundamental change of the `weight` handling as proposed in #86 ?

**Comment by williamjameshandley** [comment]
Is this issue resolved by #115?

**Comment by williamjameshandley** [comment]
> ### Simpler MWE:
> ```python
> ns = NestedSamples(root="anesthetic/tests/example_data/pc")
> fig, axes = make_2d_axes(['x0', 'x1'])
> ns.plot_2d(axes)
> ns[1:].plot_2d(axes);
> ```
> 
> ![full_vs_sliced](https://user-images.githubusercontent.com/18258042/84073700-33c09080-a9c9-11ea-8210-62844510e519.png)
> 
> This highlights how the removal of a single sample point makes the scatter plot blow up.

Running this same code now yields this plot
![Figure_1](https://user-images.githubusercontent.com/3708486/89009357-4a69d080-d304-11ea-9a2a-866f4870d551.png)

So I'm going to flag this as fixed by #115

**Comment by lukashergt** [comment]
So we basically got this for free from the new `weights` handling? Neat!
